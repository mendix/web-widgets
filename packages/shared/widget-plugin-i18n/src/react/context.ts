import { createContext } from "react";
import { PluginI18n } from "../types";

const contextPath = "com.mendix.widget.plugin.i18n.context" as const;

type I18nContext = React.Context<PluginI18n | null>;

const createI18nContext = (): I18nContext => createContext<PluginI18n | null>(null);

declare global {
    interface Window {
        [contextPath]?: ReturnType<typeof createI18nContext>;
    }
}

export function getGlobalI18nPluginContext(): I18nContext {
    return window[contextPath] ?? createI18nContext();
}
