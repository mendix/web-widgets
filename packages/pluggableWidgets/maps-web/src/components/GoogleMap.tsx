import { createElement, ReactElement, useEffect, useRef, useState } from "react";
import classNames from "classnames";
import {
    AdvancedMarker,
    APIProvider,
    InfoWindow,
    Map as GoogleMapComponent,
    MapProps,
    Pin,
    useAdvancedMarkerRef,
    useApiIsLoaded,
    useMap
} from "@vis.gl/react-google-maps";
import { Marker, SharedProps } from "../../typings/shared";
import { translateZoom } from "../utils/zoom";
import { getDimensions } from "@mendix/widget-plugin-platform/utils/get-dimensions";

export interface GoogleMapsProps extends SharedProps {
    mapId: string;
    streetViewControl: boolean;
    mapTypeControl: boolean;
    fullscreenControl: boolean;
    rotateControl: boolean;
}

export function GoogleMapContainer(props: GoogleMapsProps): ReactElement {
    return (
        <APIProvider apiKey={props.mapsToken ?? ""}>
            <GoogleMap {...props} />
        </APIProvider>
    );
}

function GoogleMap(props: GoogleMapsProps): ReactElement {
    const map = useMap();
    const isLoaded = useApiIsLoaded();
    const center = useRef<google.maps.LatLngLiteral>({
        lat: 51.906688,
        lng: 4.48837
    });
    const {
        autoZoom,
        className,
        currentLocation,
        fullscreenControl,
        locations,
        mapTypeControl,
        optionZoomControl: zoomControl,
        optionScroll: scrollwheel,
        optionDrag: draggable,
        rotateControl,
        streetViewControl,
        style,
        zoomLevel
    } = props;

    useEffect(() => {
        if (map) {
            const bounds = new google.maps.LatLngBounds();
            locations
                .concat(currentLocation ? [currentLocation] : [])
                .filter(m => !!m)
                .forEach(marker => {
                    bounds.extend({
                        lat: marker.latitude,
                        lng: marker.longitude
                    });
                });
            if (bounds.isEmpty()) {
                bounds.extend(center.current);
            }
            if (autoZoom) {
                map.fitBounds(bounds);
            } else {
                map.setCenter(bounds.getCenter());
            }
        }
    }, [map, locations, currentLocation, autoZoom]);

    const mapOptions: MapProps = {
        className: "widget-google-maps",
        defaultCenter: center.current,
        defaultZoom: autoZoom ? translateZoom("city") : zoomLevel,
        fullscreenControl,
        gestureHandling: draggable ? "auto" : "none",
        mapId: props.mapId,
        mapTypeControl,
        maxZoom: 20,
        minZoom: 1,
        rotateControl,
        scrollwheel,
        streetViewControl,
        zoomControl
    };

    return (
        <div className={classNames("widget-maps", className)} style={{ ...style, ...getDimensions(props) }}>
            <div className="widget-google-maps-wrapper">
                {isLoaded ? (
                    <GoogleMapComponent {...mapOptions}>
                        {locations
                            .concat(currentLocation ? [currentLocation] : [])
                            .filter(m => !!m)
                            .map(marker => (
                                <GoogleMapsMarker
                                    key={`marker_${marker.id ?? marker.latitude + "_" + marker.longitude}`}
                                    {...marker}
                                />
                            ))}
                    </GoogleMapComponent>
                ) : (
                    <div className="spinner" />
                )}
            </div>
        </div>
    );
}

function GoogleMapsMarker(marker: Marker): ReactElement {
    const [markerRef, googleMarker] = useAdvancedMarkerRef();
    const [infowindowShown, setInfowindowShown] = useState(false);

    const toggleInfoWindow = (): void => setInfowindowShown(previousState => !previousState);
    const closeInfoWindow = (): void => setInfowindowShown(false);

    return (
        <AdvancedMarker
            position={{
                lat: marker.latitude,
                lng: marker.longitude
            }}
            title={marker.title}
            ref={markerRef}
            onClick={() => {
                if (marker.title) {
                    toggleInfoWindow();
                }

                if (marker.onClick) {
                    marker.onClick();
                }
            }}
        >
            {infowindowShown && (
                <InfoWindow anchor={googleMarker} onCloseClick={closeInfoWindow}>
                    <span style={{ cursor: marker.onClick ? "pointer" : "none" }} onClick={marker.onClick}>
                        {marker.title}
                    </span>
                </InfoWindow>
            )}
            {marker.url && <img src={marker.url} />}
            {!marker.url && <Pin />}
        </AdvancedMarker>
    );
}
