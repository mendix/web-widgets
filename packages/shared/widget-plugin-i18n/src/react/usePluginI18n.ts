import { useState } from "react";
import { I18nController } from "../controllers/I18nController";
import { PluginI18n } from "../types";

export function usePluginI18n(languageCode: string): PluginI18n | null {
    const [plugin, setPlugin] = useState<PluginI18n | null>(() => {
        new I18nController({
            languageCode,
            onPluginChange: plugin => setPlugin(plugin)
        });

        return null;
    });

    return plugin;
}
