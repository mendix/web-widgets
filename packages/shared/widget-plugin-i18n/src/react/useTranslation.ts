import { useContext } from "react";
import { PluginI18n } from "../types";
import { getGlobalI18nPluginContext } from "./context";

export function useTranslation(): PluginI18n | null {
    return useContext(getGlobalI18nPluginContext());
}
