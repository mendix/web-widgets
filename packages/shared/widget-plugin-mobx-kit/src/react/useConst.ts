import { useRef } from "react";

export function useConst<T>(fn: T): T extends (...args: never) => infer R ? R : T {
    return (useRef<T | null>(null).current ??= typeof fn === "function" ? fn() : fn);
}
