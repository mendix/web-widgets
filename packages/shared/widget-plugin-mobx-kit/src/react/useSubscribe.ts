import { IReactionOptions, reaction } from "mobx";
import { useEffect, useState } from "react";

export function useSubscribe<Store extends object, T, FireImmediately extends boolean = false>(
    init: Store | (() => Store),
    selector: (store: Store) => T,
    options?: IReactionOptions<T, FireImmediately>
): [T, Store] {
    const [store] = useState(() => (typeof init === "function" ? init() : init));
    const [value, setValue] = useState(() => selector(store));

    function setup(): () => void {
        return reaction(() => selector(store), setValue, options);
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
    useEffect(setup, []);

    return [value, store];
}
