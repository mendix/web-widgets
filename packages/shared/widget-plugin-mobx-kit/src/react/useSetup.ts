import { useEffect } from "react";
import { useConst } from "./useConst";

export function useSetup<
    T extends {
        setup(): void | (() => void);
    }
>(fn: () => T): T {
    const obj = useConst(fn);
    useEffect(() => obj.setup(), [obj]);
    return obj;
}
