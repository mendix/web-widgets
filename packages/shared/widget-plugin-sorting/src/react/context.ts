import { useConst } from "@mendix/widget-plugin-mobx-kit/react/useConst";
import { generateUUID } from "@mendix/widget-plugin-platform/framework/generate-uuid";
import { Context, createContext, useContext, useEffect } from "react";
import { SortStoreHost } from "../stores/SortStoreHost";
import { Result, error, value } from "./result-meta";

export interface SortAPI {
    version: 1;
    host: SortStoreHost;
}

const SORT_PATH = "com.mendix.widgets.web.sortable.sortContext";

export function getGlobalSortContext(): Context<SortAPI | null> {
    const scope = window.top === window ? window : window.top;
    return ((scope as any)[SORT_PATH] ??= createContext<SortAPI | null>(null));
}

export function useSortAPI(): Result<SortAPI, Error> {
    const api = useContext(getGlobalSortContext());
    if (api === null) {
        return error(new Error("Error: widget is out of context. Please place the widget inside the Gallery header."));
    }
    return value(api);
}

export function useLockSortAPI(api: SortAPI): Result<SortAPI, Error> {
    const id = useLock(api);

    if (api.host.usedBy !== id) {
        return error(
            new Error(
                `Error: Sort API is already in use by another widget. Remove other sort widgets and refresh the page.`
            )
        );
    }

    return value(api);
}

function useLock({ host }: SortAPI): string {
    const [unlock, id] = useConst(() => {
        const id = `useLock@${generateUUID()}`;
        return [host.lock(id), id] as const;
    });

    useEffect(() => unlock, [unlock]);

    return id;
}
