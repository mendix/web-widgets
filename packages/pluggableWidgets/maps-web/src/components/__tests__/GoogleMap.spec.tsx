import "@testing-library/jest-dom";
import { render, RenderResult } from "@testing-library/react";
import { GoogleMapContainer, GoogleMapsProps } from "../GoogleMap";
import { initialize } from "@googlemaps/jest-mocks";

describe("Google maps", () => {
    const defaultProps: GoogleMapsProps = {
        autoZoom: true,
        className: "",
        currentLocation: undefined,
        fullscreenControl: false,
        height: 75,
        heightUnit: "pixels",
        locations: [],
        mapsToken: "",
        mapId: "DEMO_MAP_ID",
        mapTypeControl: false,
        optionDrag: true,
        optionScroll: true,
        optionZoomControl: true,
        rotateControl: false,
        showCurrentLocation: false,
        streetViewControl: false,
        style: {},
        width: 50,
        widthUnit: "percentage",
        zoomLevel: 10
    };

    beforeEach(() => {
        initialize();
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    function renderGoogleMap(props: Partial<GoogleMapsProps> = {}): RenderResult {
        return render(<GoogleMapContainer {...defaultProps} {...props} />);
    }

    it("renders a map with right structure", () => {
        const { asFragment } = renderGoogleMap({ heightUnit: "percentageOfWidth", widthUnit: "pixels" });
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders a map with pixels renders structure correctly", () => {
        const { asFragment } = renderGoogleMap({ heightUnit: "pixels", widthUnit: "pixels" });
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders a map with percentage of width and height units renders the structure correctly", () => {
        const { asFragment } = renderGoogleMap({ heightUnit: "percentageOfWidth", widthUnit: "percentage" });
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders a map with percentage of parent units renders the structure correctly", () => {
        const { asFragment } = renderGoogleMap({ heightUnit: "percentageOfParent", widthUnit: "percentage" });
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders a map with markers", () => {
        const { asFragment } = renderGoogleMap({
            locations: [
                {
                    title: "Mendix HQ",
                    latitude: 51.906688,
                    longitude: 4.48837,
                    url: "image:url"
                },
                {
                    title: "Gementee Rotterdam",
                    latitude: 51.922823,
                    longitude: 4.479632,
                    url: "image:url"
                }
            ]
        });
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders a map with current location", () => {
        const { asFragment } = renderGoogleMap({
            showCurrentLocation: true,
            currentLocation: {
                latitude: 51.906688,
                longitude: 4.48837,
                url: "image:url"
            }
        });
        expect(asFragment()).toMatchSnapshot();
    });
});
