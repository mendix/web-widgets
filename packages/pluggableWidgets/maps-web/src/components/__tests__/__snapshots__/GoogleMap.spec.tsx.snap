// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Google maps renders a map with current location 1`] = `
<DocumentFragment>
  <div
    class="widget-maps"
    style="width: 50%; height: 75px;"
  >
    <div
      class="widget-google-maps-wrapper"
    >
      <div
        class="widget-google-maps"
        data-testid="map"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Google maps renders a map with markers 1`] = `
<DocumentFragment>
  <div
    class="widget-maps"
    style="width: 50%; height: 75px;"
  >
    <div
      class="widget-google-maps-wrapper"
    >
      <div
        class="widget-google-maps"
        data-testid="map"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Google maps renders a map with percentage of parent units renders the structure correctly 1`] = `
<DocumentFragment>
  <div
    class="widget-maps"
    style="width: 50%; height: 75%;"
  >
    <div
      class="widget-google-maps-wrapper"
    >
      <div
        class="widget-google-maps"
        data-testid="map"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Google maps renders a map with percentage of width and height units renders the structure correctly 1`] = `
<DocumentFragment>
  <div
    class="widget-maps"
    style="width: 50%; height: auto; padding-bottom: 37.5%;"
  >
    <div
      class="widget-google-maps-wrapper"
    >
      <div
        class="widget-google-maps"
        data-testid="map"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Google maps renders a map with pixels renders structure correctly 1`] = `
<DocumentFragment>
  <div
    class="widget-maps"
    style="width: 50px; height: 75px;"
  >
    <div
      class="widget-google-maps-wrapper"
    >
      <div
        class="widget-google-maps"
        data-testid="map"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Google maps renders a map with right structure 1`] = `
<DocumentFragment>
  <div
    class="widget-maps"
    style="width: 50px; height: 37.5px;"
  >
    <div
      class="widget-google-maps-wrapper"
    >
      <div
        class="widget-google-maps"
        data-testid="map"
      />
    </div>
  </div>
</DocumentFragment>
`;
